[RPC][GUI] Adds encryption to the wallet txdata and addressbook

mxaddict opened the pull request:
closes #600

navbuilder commented:
A new build of 95a8bf3b9b38847d12b71768e4aa918dc324fcd6 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

mxaddict commented:
I've implemented the wallet encryption support

Still missing the ability to encrypt the wallet "if it was not encrypted on creation"

Currently only way to encrypt the wallet is when you create it.

mxaddict commented:
![Capture](https://user-images.githubusercontent.com/1060905/84093939-2db0cb00-aa2e-11ea-90c2-09b4c724d37a.PNG)
![Capture2](https://user-images.githubusercontent.com/1060905/84093945-2ee1f800-aa2e-11ea-8be8-707244f4d208.PNG)

mxaddict commented:
if running via cli:
```
./src/navcoind -devnet -ntpminmeasures=-1 -wallet=wallet.encrypted2.dat
PIN/PASS:****
```

or pass as an argument:
```
./src/navcoind -devnet -ntpminmeasures=-1 -wallet=wallet.encrypted2.dat -pin=test
```

navbuilder commented:
A new build of bebb30720350671b7080328c54cc7f3ec0cdfb5f has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

mxaddict commented:
Ways to encrypt the wallet:

1. Via cli on creation of wallet, just add `-pin=thepinorpasswordthatyouwant`
1. RPC call to `encrypttxdata thepinorpasswordthatyouwant`

On loading an encrypted wallet database:
1. Via cli `navcoind` or `navcoin-qt` by passing `-pin=thepasswordorpinyouusedwhenencrypting`
1. Via cli unlock by it will ask you for the password or pin you used when encrypting
1. Via GUI unlock by it will ask you for the password or pin you used when encrypting

Changing the encryption key for wallet database can also be done via `encrypttxdata newpassword` RPC call

mxaddict commented:
Tested scenarios:

* [ ] Scenario 1
1. Create wallet with pin
1. Close navcoin
1. Re-open with pin

* [ ] Scenario 2
1. Create plain wallet
1. Encrypt via rpc call `encrypttxdata`
1. Re-open with pin

* [ ] Scenario 3
1. Create wallet
1. Encrypt wallet keys (Original encryption)
1. Encrypt wallet database (`encrypttxdata`)
1. Re-open with pin
1. Spend some coins with password

* [ ] Scenario 4
1. Create wallet with pin
1. Encrypt wallet keys (Original encryption)
1. Re-open with pin
1. Spend some coins with password

* [ ] Scenario 5
1. Create plain wallet
1. Create pin
1. Create password
1. Change pin
1. Change password
1. Open wallet and spend coins

navbuilder commented:
A new build of f1e9024c2f49201b7a93a0f12a49a7a66e9d6727 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

mxaddict commented:
Some testing I did with an old wallet.dat file:

File was 1.8MB in size, after I encrypted it will a pin (Which basically updated it to latest libdb format as well) It was only 1.5MB and was working fine.

navbuilder commented:
A new build of bcd503c290bf657b0733fbf3ac2ca4aca057a222 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

navbuilder commented:
A new build of 0503da908b8e4a023763dd9e9f4f5d789d77974e has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

navbuilder commented:
A new build of fe5839dae8d10a896c1f6e60e18120107c74d5c9 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

chasingkirkjufell commented:
How do I encrypt an old wallet in qt? Does the wallet needs to be encrypted first so then encryption of txdata and addressbook can be added?

mxaddict commented:
> How do I encrypt an old wallet in qt? Does the wallet needs to be encrypted first so then encryption of txdata and addressbook can be added?

In the current state of this branch there is no limit in the order of encryption.

You can run `encrypttxda` or `encryptwallet` in any order.

chasingkirkjufell commented:
But i can't seem to find a way to encrypt it in the qt wallet. would you consider adding an in wallet button for it for qt?

mxaddict commented:
> But i can't seem to find a way to encrypt it in the qt wallet. would you consider adding an in wallet button for it for qt?

Yeah, currently no way to encrypt via the GUI (I was just running encrypttxdata from console)

I'll add it after I get some food 🤣

navbuilder commented:
A new build of 1369e0ffdc66ab8ead39f063e37e4daa0788e703 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

mxaddict commented:
@chasingkirkjufell I've added the GUI support for adding or chaning PIN

And added it to the test scenarios I have listed above.

mxaddict commented:
![Capture](https://user-images.githubusercontent.com/1060905/84577911-e42bfb80-adf2-11ea-9884-ce09f961e06a.PNG)

navbuilder commented:
A new build of e2a33362dd0243bf5df113922e23b79c403cfaf1 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

chasingkirkjufell commented:
given that the password for tx encryption is another set of password on top of the wallet itself, do you guys think we still need a minimum length requirement? it reminds me of having a quick look into current balance info of credit cards which often times only requires a 4 digit pin.

mxaddict commented:
> given that the password for tx encryption is another set of password on top of the wallet itself, do you guys think we still need a minimum length requirement? it reminds me of having a quick look into current balance info of credit cards which often times only requires a 4 digit pin.

In the current state, it required length of 1 character.

chasingkirkjufell commented:
> 
> In the current state, it only requires 1 character.

I see. I guess I was reading it off the message here.
![image](https://user-images.githubusercontent.com/38635377/84602152-157bea00-ae3a-11ea-8eda-ef65483c04f9.png)

mxaddict commented:
> > In the current state, it only requires 1 character.
> 
> I see. I guess I was reading it off the message here.
> ![image](https://user-images.githubusercontent.com/38635377/84602152-157bea00-ae3a-11ea-8eda-ef65483c04f9.png)

Yes, I just copied the same format for the suggestion text, the passphrase behaves the same way, it tells you to use 10 or more, but it only required 1.

chasingkirkjufell commented:
> > > In the current state, it only requires 1 character.
> > 
> > 
> > I see. I guess I was reading it off the message here.
> > ![image](https://user-images.githubusercontent.com/38635377/84602152-157bea00-ae3a-11ea-8eda-ef65483c04f9.png)
> 
> Yes, I just copied the same format for the suggestion text, the passphrase behaves the same way, it tells you to use 10 or more, but it only required 1.

In that case, can we add "for higher security" at the end if the suggestion sentence? 

So something like:
"Please use a pin of ten or more random characters, or eight or more words for higher security."

chasingkirkjufell commented:
Tested scenarios:

 Scenario 1

- [ ] Client starts with an old wallet
- [ ] Encrypt old wallet
- [ ] Re-open with pin

 Scenario 2

- [ ] Client starts with no wallet.dat
- [ ] Encrypt the wallet
- [ ] Re-open with pin

 Scenario 3

- [ ] Client starts with no wallet.dat
- [ ] Encrypt the wallet
- [ ] Change the encryption of the wallet
- [ ] Re-open with pin and confirm old pin would fail

mxaddict commented:
@chasingkirkjufell I've updated the message

navbuilder commented:
A new build of 32f07eafeebd718d49e4f6d11a05ff96cc85daee has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

mxaddict commented on src/qt/askpassphrasedialog.cpp:
```diff
@@ -117,53 +122,104 @@ void AskPassphraseDialog::accept()
+                                "<br><br><b>" +
+                                tr("IMPORTANT: Any previous backups you have made of your wallet file "
+                                    "should be replaced with the newly generated, encrypted wallet file. "
+                                    "For security reasons, previous backups of the unencrypted wallet file "
```
This is no longer applicable "will become useless as soon as you start using the new, encrypted wallet"

Since we disabled that feature to regenerate the master

mxaddict commented on src/qt/askpassphrasedialog.cpp:
```diff
@@ -117,53 +122,104 @@ void AskPassphraseDialog::accept()
+                                "<br><br><b>" +
+                                tr("IMPORTANT: Any previous backups you have made of your wallet file "
+                                    "should be replaced with the newly generated, encrypted wallet file. "
+                                    "For security reasons, previous backups of the unencrypted wallet file "
```
I think we can remove: "previous backups of the unencrypted wallet file will become useless as soon as you start using the new, encrypted wallet."

mxaddict commented on src/qt/askpassphrasedialog.cpp:
```diff
@@ -117,53 +122,104 @@ void AskPassphraseDialog::accept()
+                break;
+            }
+            QMessageBox::StandardButton retval = QMessageBox::question(this, tr("Confirm wallet txdata encryption"),
+                    tr("Warning: If you encrypt your wallet txdata and lose your pin, you will <b>LOSE ALL OF YOUR NAVCOINS</b>!") + "<br><br>" + tr("Are you sure you wish to encrypt your wallet txdata?"),
```
👍

aguycalled commented:
any update on this? @mxaddict

navbuilder commented:
A new build of 32f07eafeebd718d49e4f6d11a05ff96cc85daee has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

mxaddict commented:
@chasingkirkjufell I fixed the crash that was happening when clicking the ok button.

aparently it was because the thread that I created the dialog in was closed before the GUI could render changes to the button when clicking.

I fixed it by creating a threadsafe input dialog.

navbuilder commented:
A new build of 16c21835e4041507d7dd93f9705efc8076208138 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

aguycalled commented:
running `./navcoin-5.0.1/bin/navcoind -devnet -daemon` with an already encrypted wallet should not start and prompt the user to run it with `-pin=` instead.
right now the process gets locked

mxaddict commented:
Gotcha, when running without -daemon it asks for the PIN correct?

On Thu, Nov 12, 2020, 19:35 alex v. <notifications@github.com> wrote:

> running ./navcoin-5.0.1/bin/navcoind -devnet -daemon with an already
> encrypted wallet should not start and prompt the user to run it with -pin=
> instead.
> right now the process gets locked
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/navcoin/navcoin-core/pull/709#issuecomment-726023597>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDAKPONDZZCHBGMGCIEXLSPPCALANCNFSM4NXJPG5Q>
> .
>

aguycalled commented:
yes, it works

navbuilder commented:
A new build of 8cd7c824a08f0c4900ddeb2ad753ab843c3b2bd7 has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

navbuilder commented:
A new build of 93654a164a22f0046739921535002ec40f683d9f has completed succesfully!
Binaries available at https://build.nav.community/binaries/rpc-encrypt-txdata

chasingkirkjufell commented:
one last suggestion, when the user fail to put in the correct pin to unlock the wallet, can you direct the user to the asking window again instead of closing the client?
So when user sees this and clicks ok
![image](https://user-images.githubusercontent.com/38635377/99028332-5af52e00-2524-11eb-84ab-63fdafb02e46.png)
then instead of closing the wallet, it goes to this window asking for input again
![image](https://user-images.githubusercontent.com/38635377/99028369-6f392b00-2524-11eb-90d4-25ead1a1bca3.png)

mxaddict commented:
I was originally planning on doing that, but the way Berkeley DB handles
the environment on load does not allow for an easy to accomplish this.

I'm sure it's possible to do, just not sure it's worth the effort.

On Fri, Nov 13, 2020, 12:20 chasingkirkjufell <notifications@github.com>
wrote:

> one last suggestion, when the user fail to put in the correct pin to
> unlock the wallet, can you direct the user to the asking window again
> instead of closing the client?
> So when user sees this and clicks ok
> [image: image]
> <https://user-images.githubusercontent.com/38635377/99028332-5af52e00-2524-11eb-84ab-63fdafb02e46.png>
> then instead of closing the wallet, it goes to this window asking for
> input again
> [image: image]
> <https://user-images.githubusercontent.com/38635377/99028369-6f392b00-2524-11eb-90d4-25ead1a1bca3.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/navcoin/navcoin-core/pull/709#issuecomment-726498169>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIDAKJMI5SFX56GKVVYVMLSPSXYPANCNFSM4NXJPG5Q>
> .
>

chasingkirkjufell commented:
I see. In that case, since it's a minor annoyance, it's fine leaving it for now.
